pipenv/cli/options.py

```diff
@@ -0,0 +1,385 @@
+# -*- coding=utf-8 -*-
+from __future__ import absolute_import
+
+import os
+
+from click import BOOL as click_booltype
+from click import (
+    BadParameter, Group, Option, argument, echo, make_pass_decorator, option
+)
+
+from .. import environments
+from ..utils import is_valid_url
+
+
+CONTEXT_SETTINGS = dict(help_option_names=["-h", "--help"])
+
+
+class PipenvGroup(Group):
+    """Custom Group class provides formatted main help"""
+
+    def get_help_option(self, ctx):
+        from ..core import format_help
+
+        """Override for showing formatted main help via --help and -h options"""
+        help_options = self.get_help_option_names(ctx)
+        if not help_options or not self.add_help_option:
+            return
+
+        def show_help(ctx, param, value):
+            if value and not ctx.resilient_parsing:
+                if not ctx.invoked_subcommand:
+                    # legit main help
+                    echo(format_help(ctx.get_help()))
+                else:
+                    # legit sub-command help
+                    echo(ctx.get_help(), color=ctx.color)
+                ctx.exit()
+
+        return Option(
+            help_options,
+            is_flag=True,
+            is_eager=True,
+            expose_value=False,
+            callback=show_help,
+            help="Show this message and exit.",
+        )
+
+
+class State(object):
+    def __init__(self):
+        self.index = None
+        self.extra_index_urls = []
+        self.verbose = False
+        self.pypi_mirror = None
+        self.python = None
+        self.two = None
+        self.three = None
+        self.site_packages = False
+        self.clear = False
+        self.system = False
+        self.installstate = InstallState()
+
+
+class InstallState(object):
+    def __init__(self):
+        self.dev = False
+        self.pre = False
+        self.selective_upgrade = False
+        self.keep_outdated = False
+        self.skip_lock = False
+        self.ignore_pipfile = False
+        self.sequential = False
+        self.code = False
+        self.requirementstxt = None
+        self.deploy = False
+        self.packages = []
+        self.editables = []
```

Comment:
Why not use attrs for this? I’m not its biggest fan, but this is almost as good a use case as it gets.

Reply:
Not really...  there is never any I initialization or validation here. Literally just putting a constrained list of things. I took the implementation directly from Armin. There is no reason to complicate it. There is absolutely nothing happening in this class besides initialization. Attrs adds 0 value